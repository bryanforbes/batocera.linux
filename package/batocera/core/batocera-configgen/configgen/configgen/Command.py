from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
