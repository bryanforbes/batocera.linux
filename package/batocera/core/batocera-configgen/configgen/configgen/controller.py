from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
