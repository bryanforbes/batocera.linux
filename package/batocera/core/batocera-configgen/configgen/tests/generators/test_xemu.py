from __future__ import annotations

import filecmp
from pathlib import Path
from typing import TYPE_CHECKING, Any

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS, SAVES
from configgen.config import SystemConfig
from configgen.generators.xemu.xemuGenerator import XemuGenerator
from tests.generators.base import GeneratorBaseTest
from tests.mock_controllers import make_player_controller_list

if TYPE_CHECKING:
    from pyfakefs.fake_filesystem import FakeFilesystem
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controller, Controllers
    from configgen.Emulator import Emulator


class TestXemuGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[XemuGenerator]:
        return XemuGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'xbox'

    @pytest.fixture
    def emulator(self) -> str:
        return 'xemu'

    @pytest.fixture(autouse=True)
    def fs(self, fs: FakeFilesystem) -> FakeFilesystem:
        fs.create_file('/usr/share/xemu/data/xbox_hdd.qcow2', contents='xbox_hdd.qcow2')
        return fs

    @pytest.mark.parametrize(
        ('mock_system_config', 'result'),
        [
            ({}, 4 / 3),
            ({'xemu_scaling': 'scale'}, 4 / 3),
            ({'xemu_aspect': 'native'}, 4 / 3),
            ({'xemu_scaling': 'stretch'}, 16 / 9),
            ({'xemu_aspect': '16x9'}, 16 / 9),
            ({'xemu_scaling': 'scale', 'xemu_aspect': '16x9'}, 16 / 9),
        ],
        ids=str,
    )
    def test_get_in_game_ratio(  # pyright: ignore
        self, generator: XemuGenerator, mock_system_config: dict[str, Any], result: bool
    ) -> None:
        assert generator.getInGameRatio(SystemConfig(mock_system_config), {'width': 0, 'height': 0}, Path()) == result

    def test_generate(
        self,
        generator: XemuGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'xbox' / 'rom.iso',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert filecmp.cmp('/usr/share/xemu/data/xbox_hdd.qcow2', SAVES / 'xbox' / 'xbox_hdd.qcow2')
        assert (CONFIGS / 'xemu' / 'xemu.toml').read_text() == snapshot(name='config')

    def test_generate_existing(
        self,
        generator: XemuGenerator,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file(SAVES / 'xbox' / 'xbox_hdd.qcow2', contents='saved xbox_hdd.qcow2')
        fs.create_file(
            CONFIGS / 'xemu' / 'xemu.toml',
            contents="""[general]
skip_boot_anim = true

[sys]
mem_limit = "128"

[sys.files]
flashrom_path = "/tmp/Complex_4627.bin"

[audio]
use_dsp = true

[display]
renderer = "VULKAN"

[display.quality]
surface_scale = 2

[display.window]
fullscreen_on_startup = false

[display.ui]
show_menubar = true

[input.bindings]
port1 = "foobarbaz"

[net]
enable = true

[net.udp]
remote_addr = "10.10.10.10"
bind_addr = "10.10.10.10"
""",
        )

        generator.generate(
            mock_system,
            ROMS / 'xbox' / 'rom.iso',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (SAVES / 'xbox' / 'xbox_hdd.qcow2').read_text() == 'saved xbox_hdd.qcow2'
        assert (CONFIGS / 'xemu' / 'xemu.toml').read_text() == snapshot

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'xemu_bootanim': 'true'},
            {'xemu_memory': '128'},
            {'xemu_use_dsp': 'true'},
            {'xemu_api': 'OPENGL'},
            {'xemu_render': '2'},
            {'xemu_vsync': 'false'},
            {'xemu_scaling': 'stretch'},
            {'xemu_aspect': '16x9'},
            {'xemu_networktype': 'nat'},
            {'xemu_udpremote': '10.10.10.10'},
            {'xemu_udpbind': '10.10.10.10'},
        ],
        ids=str,
    )
    def test_generate_config(
        self,
        generator: XemuGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'xbox' / 'rom.iso',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'xemu' / 'xemu.toml').read_text() == snapshot

    @pytest.mark.system_name('chihiro')
    def test_generate_chihiro(
        self,
        generator: XemuGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'chihiro' / 'rom.iso',
            one_player_controllers,
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'xemu' / 'xemu.toml').read_text() == snapshot

    def test_generate_controllers(
        self,
        generator: XemuGenerator,
        mock_system: Emulator,
        generic_xbox_pad: Controller,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'xbox' / 'rom.iso',
            make_player_controller_list(
                generic_xbox_pad, generic_xbox_pad, generic_xbox_pad, generic_xbox_pad, generic_xbox_pad
            ),
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )

        assert (CONFIGS / 'xemu' / 'xemu.toml').read_text() == snapshot
