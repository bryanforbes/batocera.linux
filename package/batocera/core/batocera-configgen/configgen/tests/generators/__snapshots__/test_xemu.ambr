# serializer version: 1
# name: TestXemuGenerator.test_generate
  tuple(
    list([
      '/usr/bin/xemu',
      '-config_path',
      '/userdata/system/configs/xemu/xemu.toml',
    ]),
    dict({
      'SDL_GAMECONTROLLERCONFIG': '030000005e0400000a0b000005040000,Generic Xbox pad,platform:Linux,b:b0,a:b1,dpdown:h0.4,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:b6,leftstick:b11,dpleft:h0.8,rightshoulder:b5,leftshoulder:b4,righttrigger:b7,rightstick:b12,dpright:h0.2,back:b8,start:b9,dpup:h0.1,y:b2,x:b3,',
      'XDG_CONFIG_HOME': '/userdata/system/configs',
    }),
  )
# ---
# name: TestXemuGenerator.test_generate[config]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_chihiro
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "128"
  
  [sys.files]
  flashrom_path = "/userdata/bios/cerbios.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/chihiro/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "OPENGL"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_api': 'OPENGL'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "OPENGL"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_aspect': '16x9'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "16x9"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_bootanim': 'true'}]
  '''
  [general]
  skip_boot_anim = true
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_memory': '128'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "128"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_networktype': 'nat'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = true
  backend = "nat"
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_render': '2'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 2
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_scaling': 'stretch'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "stretch"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_udpbind': '10.10.10.10'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  bind_addr = "10.10.10.10"
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_udpremote': '10.10.10.10'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  remote_addr = "10.10.10.10"
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_use_dsp': 'true'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = true
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_config[{'xemu_vsync': 'false'}]
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = false
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_controllers
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  port2 = "030000005e0400000a0b000005040000"
  port3 = "030000005e0400000a0b000005040000"
  port4 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  
  
  '''
# ---
# name: TestXemuGenerator.test_generate_existing
  '''
  [general]
  skip_boot_anim = false
  show_welcome = false
  screenshot_dir = "/userdata/screenshots"
  
  [sys]
  mem_limit = "64"
  
  [sys.files]
  flashrom_path = "/userdata/bios/Complex_4627.bin"
  bootrom_path = "/userdata/bios/mcpx_1.0.bin"
  hdd_path = "/userdata/saves/xbox/xbox_hdd.qcow2"
  eeprom_path = "/userdata/saves/xbox/xemu_eeprom.bin"
  dvd_path = "/userdata/roms/xbox/rom.iso"
  
  [audio]
  use_dsp = false
  
  [display]
  renderer = "VULKAN"
  
  [display.quality]
  surface_scale = 1
  
  [display.window]
  fullscreen_on_startup = true
  startup_size = "1920x1080"
  vsync = true
  
  [display.ui]
  show_menubar = false
  fit = "scale"
  aspect_ratio = "auto"
  
  [input.bindings]
  port1 = "030000005e0400000a0b000005040000"
  
  [net]
  enable = false
  
  [net.udp]
  remote_addr = "10.10.10.10"
  bind_addr = "10.10.10.10"
  
  
  '''
# ---
# name: TestXemuGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
    }),
    'name': 'xemu',
  })
# ---
